docs/contrib/DeveloperGuide.md

```diff
@@ -0,0 +1,352 @@
+
+``` cpp
+    Tool *makeTool();
+    void initTool(OpenRoad *openroad);
+    void deleteTool(Tool *tool);
+```
+
+The OpenRoad::init() function calls all of the makeTool functions and
+then all of the initTool() functions. The init functions are called from
+the bottom of the tool dependences. Each init function grabs the state
+it needs out of the OpenRoad instance.
+
+## Commands
+
+Tools should provide Tcl commands to control them. Tcl object based tool
+interfaces are not user friendly. Define Tcl procedures that take
+keyword arguments that reference the OpenRoad object to get tool state.
+OpenSTA has Tcl utilities to parse keyword arguements
+(`sta::parse_keyword_args`). See `OpenSTA/tcl/*.tcl` for
+examples. Use swig to define internal functions to C++ functionality.
+
+Tcl files can be included by encoding them in cmake into a string that
+is evaluated at run time (See Resizer::init()).
+
+## Errors
+
+Tools should report errors to the user using the `ord::error` function
+defined in `include/openroad/Error.hh`. `ord::error` throws
+`ord::Exception`. The variables `ord::exit_on_error` and
+`ord::file_continue_on_error` control how the error is handled. If
+`ord::exit_on_error` is `true` OpenROAD reports the error and exits. If
+the error is encountered while reading a file with the `source` or
+`read_sdc` commands and `ord::file_continue_on_error` is `false` no
+other commands are read from the file. The default values of both
+variables is `false`.
+
+## Test
+
+Each "tool" has a /test directory containing a script nameed
+"regression" to run "unit" tests. With no arguments it should run
+default unit tests.
+
+No database files should be in tests. Read LEF/DEF/Verilog to make a
+database.
+
+The regression script should not depend on the current working
+directory. It should be able to be run from any directory. Use filenames
+relative to the script name rather the the current working directory.
+
+Regression scripts should print a consise summary of test failures. The
+regression script should return an exit code of zero if there are no
+errors and 1 if there are errors. The script should **not** print
+thousands of lines of internal tool info.
+
+Regression scripts should pass the `-no_init` option to openroad so that
+a user's init file is not sourced before the tests runs.
+
+Regression scripts should add output files or directories to
+`.gitignore` so that running does note leave the source repository
+"dirty".
+
+The Nangate45 open source library data used by many tests is in
+`test/Nangate45`. Use the following command to add a link in the tool
+command
+
+``` shell
+    cd tool/test
+    ln -s ../../../test/Nangate45
+```
+
+After the link is installed, the test script can read the liberty file
+with the command shown below.
+
+``` tcl
+    read_liberty Nangate45/Nangate45_typ.lib
+```
+
+## Builds
+
+Checking out the OpenROAD repo with --recursive installs all of the
+OpenRoad tools and their submodules.
+
```

Comment:
This may be appropriate place to address the dependency handling reported in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/106

Reply:
The dependencies to buidl I was planning on handle on these files:
docs/user/BuildWithDocker.md
docs/user/BuildLocally.md
They are [referenced here](https://github.com/The-OpenROAD-Project/OpenROAD/blob/bd6c310697dcb53daa239c3cb0720c4a8d0f381f/docs/user/GettingStarted.md#setup).

Reply:
Fixed on 6d5d6cd